tg-x/network-architecture.md

```diff
@@ -0,0 +1,234 @@
+Local engines communicate with each other directly,
+while they reach remote nodes via the Router and Transport engines.
+
+The Transport engine is responsible for maintaining authenticated and encrypted network connections between nodes
+via a modular transport system that supports various transport protocols with different reliability, latency, and privacy guarantees.
+Transport protocols initially supported include QUIC and TLS,
+later this can be extended with onion-routed, mix, mesh, and delay-tolerant network protocols.
+
+The Router engine routes incoming and outgoing messages from and to local engines,
+based on the destination address of each message, which consists of a node and engine identity.
+
+<figure>
+
+```mermaid
+flowchart LR
+  T((Transport))
+  R(((Router)))
+  X((Engine_X))
+  Y((Engine_Y))
+
+  X & Y --- R --- T
+```
+
+<figcaption>A node with its engines.</figcaption>
+</figure>
+
+## Communication patterns
+
+The network supports unicast, multicast, and anycast communication patterns.
+
+- *Unicast*: direct messages between two engines.
+- *Multicast*: few-to-many messages sent to a publish-subscribe topic and delivered to all subscribers.
+- *Anycast*: messages sent to any known member of a domain.
+
+<figure>
+
+```mermaid
+flowchart LR
+
+subgraph NA[Node_A]
+  AT((Transport_A))
+  AR(((Router_A)))
+  AX((Engine_AX))
+  AY((Engine_AY))
+
+  AX --> AR --> AT
+  AY -.- AR
+
+end
+
+AT --> BT
+
+subgraph NB[Node_B]
+  BT(((Transport_B)))
+  BR(((Router_B)))
+  BX((Engine_BX))
+  BY((Engine_BY))
+
+  BT --> BR --> BX
+  BY -.- BR
+end
+```
+
+<figcaption>Unicast communication: Engine AX sends a message to BX via the Router And Transport engines.</figcaption>
+</figure>
+
+## Network Architecture
+
+The network consists of nodes that may communicate with each other directly,
+and participate in any number of domains.
+
+Nodes in the core network are directly reachable via transport address(es)
+specified in their signed node advertisements.
+
+Mobile nodes of end-users may communicate directly on edge networks,
+and reach the core network via relays
+that store and forward messages for them.
+
+This results in a two-tier network architecture,
+where relays provide reliable message delivery and basic location privacy for mobile nodes.
+Mobile nodes with higher location privacy requirements
+may use an onion-routed or mix network to connect to the core network.
+
+<figure>
+
+```mermaid
+flowchart LR
+
+subgraph Core
+  A((A))
+  B((B))
+  C((C))
+  D((D))
+  R1((R1))
+  R2((R2))
+  
+  A --- B & C --- D
+  C --- R1
+  D --- R2
+end
+
+subgraph Edge_A
+  K --- L --- M & N --- O
+end
+
+subgraph Edge_B
+  P --- Q & R --- S
+end
+
+R1 -.- L & P
+R2 -.- O & R
+```
+
+<figcaption>Two-tier network architecture with core & edge networks.</figcaption>
+</figure>
+
+## Domains
+
+The network consists of sovereign domains.
+A domain is a sub-network with its own *cryptographic identity*.
```

Comment:
nodes also have crypto ids
- what's the difference?
- what are the commonalities